pkg/controller/operators/olm/operatorgroup.go

```diff
@@ -18,63 +17,137 @@ import (
 			}
 		}
 	}
 
-	for _, csv := range csvsInNamespace {
-		if err := a.copyCsvToTargetNamespace(csv, op, targetedNamespaces); err != nil {
+	a.Log.Debug("CSV annotation completed")
+
+	return nil
+}
+
+// ensureProvidedAPIClusterRole ensures that a clusterrole exists (admin, edit, or view) for a single provided API Type
+func (a *Operator) ensureProvidedAPIClusterRole(operatorGroup *v1alpha2.OperatorGroup, csv *v1alpha1.ClusterServiceVersion, namePrefix, suffix, group, resource string) error {
+	clusterRole := &rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: namePrefix + suffix,
+			Labels: map[string]string{
+				kubeRBACAggregationKeyPrefix + suffix:       "true",
+				operatorGroupAggregrationKeyPrefix + suffix: operatorGroup.GetName(),
+			},
+		},
+		Rules: []rbacv1.PolicyRule{{Verbs: VerbsForSuffix[suffix], APIGroups: []string{group}, Resources: []string{resource}}},
+	}
+	ownerutil.AddNonBlockingOwner(clusterRole, csv)
+	existingCR, err := a.OpClient.KubernetesInterface().RbacV1().ClusterRoles().Create(clusterRole)
+	if k8serrors.IsAlreadyExists(err) {
+		if existingCR != nil && reflect.DeepEqual(existingCR.Labels, clusterRole.Labels) && reflect.DeepEqual(existingCR.Rules, clusterRole.Rules) {
+			return nil
+		}
+		if _, err = a.OpClient.UpdateClusterRole(clusterRole); err != nil {
+			a.Log.WithError(err).Errorf("Update existing cluster role failed: %v", clusterRole)
 			return err
 		}
+	} else if err != nil {
+		a.Log.WithError(err).Errorf("Create cluster role failed: %v", clusterRole)
+		return err
 	}
+	return nil
+}
+
+// ensureClusterRolesForCSV ensures that ClusterRoles for writing and reading provided APIs exist for each operator
+func (a *Operator) ensureClusterRolesForCSV(csv *v1alpha1.ClusterServiceVersion, operatorGroup *v1alpha2.OperatorGroup) error {
+	for _, owned := range csv.Spec.CustomResourceDefinitions.Owned {
+		nameGroupPair := strings.SplitN(owned.Name, ".", 2) // -> etcdclusters etcd.database.coreos.com
+		if len(nameGroupPair) != 2 {
+			return fmt.Errorf("Invalid parsing of name '%v', got %v", owned.Name, nameGroupPair)
+		}
+		plural := nameGroupPair[0]
+		group := nameGroupPair[1]
+		namePrefix := fmt.Sprintf("%s-%s-", owned.Name, owned.Version)
 
-	for _, csv := range csvsInNamespace {
-		if err := a.ensureRBACInTargetNamespace(csv, op, targetedNamespaces); err != nil {
+		if err := a.ensureProvidedAPIClusterRole(operatorGroup, csv, namePrefix, AdminSuffix, group, plural); err != nil {
+			return err
+		}
+		if err := a.ensureProvidedAPIClusterRole(operatorGroup, csv, namePrefix, EditSuffix, group, plural); err != nil {
+			return err
+		}
+		if err := a.ensureProvidedAPIClusterRole(operatorGroup, csv, namePrefix, ViewSuffix, group, plural); err != nil {
 			return err
 		}
 	}
-	log.Debug("CSV annotation completed")
+	for _, owned := range csv.Spec.APIServiceDefinitions.Owned {
+		namePrefix := fmt.Sprintf("%s-%s-", owned.Name, owned.Version)
```

Comment:
Is `ownee.Name` always set for `APIServiceDefinitions`?

Reply:
It is not, but I want to review all of the current ones, add them, and then go make it `required` in the json schema